Success bar that represents successful attack (not ECONNABORTED)

taras-danyliuk opened the pull request:
Potentially resolves:
https://github.com/opengs/uashield/issues/90
https://github.com/opengs/uashield/issues/59

<img width="683" alt="image" src="https://user-images.githubusercontent.com/13213340/156811163-621260d5-b208-4bee-8f89-ce44700b19cb.png">

Andrii-L-DDoSer commented on src/pages/Index.vue:
```diff
@@ -180,6 +189,7 @@ export default defineComponent({
     const ddosEnabled = ref(true)
     const forceProxy = ref(true)
     const atackCounter = ref(0)
+    const atackErrorCounter = ref(0)
```
But change to `attackErrorCounter`, atack is typo.

taras-danyliuk replied:
@Andrii-L-DDoSer I know its typo, but everywhere in the file `atack` is used(

Andrii-L-DDoSer reviewed (commented):
Btw why u commit `yarn.lock` ? You don't have new packages. Have u executed `npm audit` or `yarn audit`?

taras-danyliuk commented:
> Btw why u commit `yarn.lock` ? You don't have new packages. Have u executed `npm audit` or `yarn audit`? To avoid conflicts, I would suggest to remove it.
> 
> cc/ @Razikus

yarn.lock is already in the repo, so much changes in this file was due node version difference. Anyway I removed that file from the PR

Razikus commented:
Hey, great work!
I need to discuss if we want really to merge it
It may be sad for someone, that only 20% requests are with status 200 
But these 20% * x * tousands of people are making job

taras-danyliuk commented:
Yeah, I though about it. The initial goal was to create some kind of indicator when VPN should be relauched for a new IP.

I can add a message related to some threshold. E.g. if success rate is lower than x% - message will show "It is recommended to use new IP" and if higher - "This IP is still useful. Keep it up and running!!!"

Razikus commented:
yep, that's good approach i will also discuss internally

taras-danyliuk commented:
@Razikus I've updated the PR. Added suggestion message based on latest 500 requests. If there will be less than 10% of success - message will suggest to change ip!

Status bar is still linked to the total amount of the requests.

<img width="888" alt="image" src="https://user-images.githubusercontent.com/13213340/156834723-bfae8275-a8b9-4568-98bc-4f04c797b6d1.png">

Andrii-L-DDoSer commented on src-worker/runner.ts:
```diff
@@ -124,7 +130,12 @@ export class Runner {
 
-        this.eventSource.emit('attack', { type: 'atack', url: target.site.page, log: `${target.site.page} | ${code}` })
+        this.eventSource.emit('attack', {
+          type: 'atack',
```
I know it's more to change, but ` type: 'atack',` => `attack` - I assume it's widely used type, so you can ignore my comment. Sure thing it's extremely low priority these days. That would be refactoring.

Andrii-L-DDoSer commented on src/i18n/ua-ua/index.ts:
```diff
@@ -27,6 +27,11 @@ export default {
+    },
+    successBar: {
+      title: 'Статистика успішних реквестів',
+      restartVpn: 'Відсоток успішних реквестів менше 10, рекомендуємо змінити IP',
```
реквестів => запитів, I would say, but реквест is also OK ;) Просто не всі користувачі будуть знати ІТ шний сленг...

Razikus commented:
Wow, really nice, thanks for this PR! 
Will review and check it tommorow

about yarn lock - is just a practice here to not commit it because someone can intentionaly inject something bad

yastasiuk commented on src/pages/Index.vue:
```diff
@@ -136,12 +155,21 @@ export default defineComponent({
+
+      // Latest 500 requests success/failure - used for IP change suggestion
+      if (this.latest500Responses.length > 500) this.latest500Responses.shift()
+      this.latest500Responses.push(data.success)
```
Хм, мне кажется для большого к-ва "тредов" может работать достаточно так себе(в случае если за раз тебе будет прилетать 500+ ответов) - да и в целом вопрос на счет UI не будет ли он глючить, например, при 1к "раннеров". Всё же с каждым изменением приходится прогонять `shift, filter`, а это +O(n) на ровном месте и иногда это вообще не нужно(для runnersNumber > 500)
Я бы тут лучше немного сменил эту логику, чтобы не использовать массивы, а вручную считать, что-то ака
```
// state
lastMinuteStatus: { success: 0, failed: 0 },
currentStatus: { success: 0, failed: 0 }
// methods
serveAtack(..) {
...
  if (data.success){
    currentStatus.successful =+ 1
    lastMinuteStatus.successful += 1
  } else {
    currentStatus.failed += 1
    lastMinuteStatus.failed += 1
  }
  ...
  },
  
  // todo: naming
checkStatus() {
 if (lastMinuteStatus.successful < lastMinuteStatus.failed) {
    // suggest IP change
    lastMinuteStatus.successful = 0
    lastMinuteStatus.failed = 0
 },
}
...
mounted() {
   setTimeout(() => { this.checkStatus() }, 60 * 1000)
}

yastasiuk replied:
*если что, то отображать статистику по `currentStatus `

taras-danyliuk replied:
First of all for bigger amount of runners: please take a look how runners are implemented. Each runner has it's own random target - so the amount of runner doesn't impact the amount of simultaneous ECONNABORTED errors.

In regards to O(n) for an array of 500 elements - it is very very very fast.

In regards to rerender (UI глючить) - we are already rerendering after each response by incrementing counter and adding log line!

Your solution - once in a minute (not even in a minute because `setTimeout(() =>, 60000)` is not a minute) we will reset everything, so if first 10 requests will be failed we will need to show the message for IP change, but all other 90 will be ok - we will need to show the message that IP is fine.
We will end up having flashing text that will be changing once in a minute and will be confusing everyone for sure!

taras-danyliuk replied:
PS: please use English in your reviews. The main contributor of this project is from Poland and obviously he doesn't understand Ukrainian 😉

yastasiuk replied:
As far as i can see `latest500Responses` is aggregated for all requests(`attack` event is bubbled to the `doser` and after that we're sending it to UI).
Therefore technically for large `n` it can be more flashy than checking it each minute(or some other arbitrary number).
But anyway if you say it works good for "large" N(i'd say 256-1k), than you can resolve this thread :)

yastasiuk commented on src/pages/Index.vue:
```diff
@@ -180,6 +208,7 @@ export default defineComponent({
     const ddosEnabled = ref(true)
     const forceProxy = ref(true)
     const atackCounter = ref(0)
+    const atackErrorCounter = ref(0)
```
`attack`

opengs commented:
@taras-danyliuk the problem is, that ECONNABORTED is also may be success :( 
ECONNABORTED raises in two cases:
1) when target server is down - SUCCESS // raises in most of the cases
2) target server has anti-ddos - ERROR

And there are no way to determine the case.

So when our attack succeeded, we will have error. So we will have misleading information :(

Also most of our users dont know wha is VPN and what is status code. So better not to make panic that something may not work as expected :D . Using VPN is the best approach, but ECONNABORTED with proxy is ok

taras-danyliuk commented:
@opengs What about making that suggestion optional and show it only if proxy is disabled.

Also can you think of any solution how we can suggest to change IP?

Razikus commented:
Maybe some kind of switch "Suggest when to  change VPN " or something

taras-danyliuk commented:
I can add that, but as @opengs mentioned ECONNABORTED response not always means that your IP was blocked.

WebManG commented:
There may be successful attempts to show in green color, and errors in red? It's familiar and userfriendly.

opengs commented:
We took all the seggestions and added this in new version. Also there was changes in attack rate calculation, so it was hard to apply this PR. Thank you for your work on this! Please check 0.0.17.
![image](https://user-images.githubusercontent.com/34212339/159806085-37e6b15a-ff72-45c7-ae5c-bddc86746f7f.png)
